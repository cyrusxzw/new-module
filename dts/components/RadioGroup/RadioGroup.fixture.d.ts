export namespace RadioGroupFixture {
  const className: string;
  const dataTestRef: string;
  const errorMessage: string;
  const name: string;
  function onChange(): any;
  const options: {
    value: string;
    label: string;
  }[];
  const value: string;
}
