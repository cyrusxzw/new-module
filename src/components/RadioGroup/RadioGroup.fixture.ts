const RadioGroupFixture = {
  className: '',
  dataTestRef: 'RADIO_GROUP_TEST_REF',
  errorMessage: 'Test Error Messsage',
  name: 'radio-group',
  onChange: () => undefined,
  options: [
    {
      value: 'morbi-id-nulla',
      label: 'Morbi id nulla',
    },
    {
      value: 'nullam-ultrices',
      label: 'Nullam ultrices',
    },
  ],
  value: 'morbi-id-nulla',
};

export { RadioGroupFixture };
