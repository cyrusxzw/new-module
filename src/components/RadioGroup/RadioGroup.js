import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './RadioGroup.module.css';

const RadioGroup = ({
  className,
  dataTestRef,
  errorMessage,
  name,
  onChange,
  options,
  theme,
  value,
}) => {
  const classSet = cx(styles.base, className);

  if (options.length === 0) {
    return null;
  }

  if (options.length === 1) {
    return (
      <>
        <div className={classSet}>
          <span className={cx(styles.singleLabel, styles[theme])}>
            {options[0].label}
          </span>
        </div>
        {errorMessage && (
          <div className={styles.errorMessage}>{errorMessage}</div>
        )}
      </>
    );
  }

  return (
    <>
      <ul className={classSet}>
        {options.map(({ label: optionLabel, value: optionValue }) => (
          <li className={styles.radio} key={optionValue}>
            <label
              className={cx(styles.label)}
              htmlFor={`option-${optionValue}`}
            >
              <input
                aria-checked={optionValue === value}
                checked={optionValue === value}
                className={styles.input}
                data-test-ref={dataTestRef}
                id={`option-${optionValue}`}
                name={name}
                onChange={onChange}
                tabIndex={0}
                type="radio"
                value={optionValue}
              />
              <span className={cx(styles.pot, styles[theme])} />
              <span className={cx(styles.labelContent, styles[theme])}>
                {optionLabel}
              </span>
            </label>
          </li>
        ))}
      </ul>
      {errorMessage && (
        <div className={styles.errorMessage}>{errorMessage}</div>
      )}
    </>
  );
};

RadioGroup.propTypes = {
  className: PropTypes.string,
  dataTestRef: PropTypes.string,
  errorMessage: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string,
    }),
  ),
  theme: PropTypes.oneOf(['dark', 'light']),
  value: PropTypes.string,
};

RadioGroup.defaultProps = {
  className: undefined,
  dataTestRef: undefined,
  errorMessage: undefined,
  name: undefined,
  onChange: undefined,
  options: [],
  theme: 'dark',
  value: undefined,
};

export { RadioGroup };
