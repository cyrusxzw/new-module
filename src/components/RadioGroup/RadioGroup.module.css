@import 'styles/index';

%user-input {
  padding: 0;
  border: 0;
  margin: 0;
  line-height: 1.4;
  outline: none;
  text-overflow: ellipsis;

  &.dark {
    color: var(--color-dark-copy);
  }

  &.light {
    color: var(--color-light-copy);
  }
}

.base {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 0;
  margin: -2px 0;
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(15px);
  font-weight: var(--font-weight-normal);
  line-height: 1.4;
  list-style: none;

  &,
  & * {
    box-sizing: border-box;
  }
}

.radio {
  padding: 0;
  margin-top: 4px;
  margin-right: 25px;
  margin-bottom: 4px;

  &:last-child {
    margin-right: 0;
  }
}

.input {
  @extend %user-input;

  position: fixed;
  width: 0;
  cursor: pointer;
  opacity: 0;
}

.label {
  cursor: pointer;
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(14px);
}

.labelContent {
  padding-left: 10px;

  &.dark {
    color: var(--color-dark-label);
  }

  &.light {
    color: var(--color-light-label);
  }

  .input:checked ~ &.dark {
    color: var(--color-grey-10);
  }

  .input:checked ~ &.light {
    color: var(--color-grey-90);
  }
}

.singleLabel {
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(14px);

  &.dark {
    color: var(--color-dark-label);
  }

  &.light {
    color: var(--color-light-copy);
  }
}

.pot {
  position: relative;
  z-index: 1;
  width: 13px;
  height: 13px;
  border-width: 1px;
  border-style: solid;
  background-color: transparent;
  border-radius: 100%;
  float: left;
  transform: translateY(3px);
  transition: all 0.5s ease;
  transition: border-color 0.2s var(--easing-ease-out-cubic);

  &.dark {
    border-color: var(--color-grey-30);
  }

  &.light {
    border-color: var(--color-grey-60);
  }

  &::after {
    position: absolute;
    top: 3px;
    right: 3px;
    bottom: 3px;
    left: 3px;
    display: block;
    background-color: transparent;
    border-radius: 100%;
    content: '';
    transition: background-color 0.2s var(--easing-ease-in-out-cubic);
  }

  .input:hover ~ &.dark::after {
    background-color: var(--color-grey-50);
  }

  .input:hover ~ &.light::after {
    background-color: var(--color-grey-65);
  }

  .input:checked ~ &.dark::after {
    background-color: var(--color-grey-30);
  }

  .input:checked ~ &.light::after {
    background-color: var(--color-grey-60);
  }
}
