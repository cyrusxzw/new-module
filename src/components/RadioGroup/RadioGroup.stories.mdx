import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import { RadioGroup } from './RadioGroup';
import { RadioGroupFixture } from './RadioGroup.fixture';

<Meta title="Components/RadioGroup" component={RadioGroup} />

# RadioGroup

<Canvas>
  <Story name="Base component">
    <State
      initialState={{
        value: '',
      }}
    >
      {(state, setState) => {
        return (
          <RadioGroup
            errorMessage={knobs.text('errorMessage', '', 'Content')}
            name={knobs.text('name', 'radioGroup', 'Content')}
            onChange={(event) => {
              event.persist();
              const {
                target: { value },
              } = event;
              setState({ ...state, value });
            }}
            options={knobs.object(
              'options',
              RadioGroupFixture.options,
              'Content',
            )}
            value={state.value}
          />
        );
      }}
    </State>
  </Story>
</Canvas>

## API

<ArgsTable of={RadioGroup} />
