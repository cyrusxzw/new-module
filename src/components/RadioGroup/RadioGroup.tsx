import React from 'react';
import cx from 'classnames';
import type { RadioGroupType } from './RadioGroup.types';
import styles from './RadioGroup.module.css';

const RadioGroup: RadioGroupType = ({
  className,
  dataTestRef,
  errorMessage,
  name,
  onChange,
  options = [],
  theme = 'dark',
  value,
}) => {
  const classSet = cx(styles.base, className);

  if (options.length === 0) {
    return null;
  }

  if (options.length === 1) {
    return (
      <>
        <div className={classSet}>
          <span className={cx(styles.singleLabel, styles[theme])}>
            {options[0].label}
          </span>
        </div>
        {errorMessage && <div>{errorMessage}</div>}
      </>
    );
  }

  return (
    <>
      <ul className={classSet}>
        {options.map(({ label: optionLabel, value: optionValue }) => (
          <li className={styles.radio} key={optionValue}>
            <label
              className={cx(styles.label)}
              htmlFor={`option-${optionValue}`}
            >
              <input
                aria-checked={optionValue === value}
                checked={optionValue === value}
                className={styles.input}
                data-test-ref={dataTestRef}
                id={`option-${optionValue}`}
                name={name}
                onChange={onChange}
                tabIndex={0}
                type="radio"
                value={optionValue}
              />
              <span className={cx(styles.pot, styles[theme])} />
              <span className={cx(styles.labelContent, styles[theme])}>
                {optionLabel}
              </span>
            </label>
          </li>
        ))}
      </ul>
      {errorMessage && <div>{errorMessage}</div>}
    </>
  );
};

export { RadioGroup };
