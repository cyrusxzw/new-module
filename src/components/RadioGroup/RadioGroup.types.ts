import type { ChangeEvent } from 'react';
import type { ComponentWithoutChildren, Themes } from '~/types';

type RadioGroupProps = {
  className?: string;
  dataTestRef?: string;
  errorMessage?: string;
  name: string;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  options?: {
    label?: string;
    value?: string;
  }[];
  theme?: Themes;
  value?: string;
};

type RadioGroupType = ComponentWithoutChildren<RadioGroupProps>;

export { RadioGroupType };
