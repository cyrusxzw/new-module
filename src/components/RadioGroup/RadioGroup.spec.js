import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe } from 'jest-axe';
import { RadioGroup } from './RadioGroup';
import { RadioGroupFixture } from './RadioGroup.fixture';

describe('<RadioGroup />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <RadioGroup
        className={RadioGroupFixture.className}
        dataTestRef={RadioGroupFixture.dataTestRef}
        errorMessage={RadioGroupFixture.errorMessage}
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={RadioGroupFixture.options}
        value={RadioGroupFixture.value}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should return correct number of items', () => {
    const itemCount = RadioGroupFixture.options.length;

    render(
      <RadioGroup
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={RadioGroupFixture.options}
      />,
    );

    const list = screen.getByRole('list');
    const items = screen.getAllByRole('listitem');

    expect(list).toBeInTheDocument();
    expect(items).toHaveLength(itemCount);
  });

  it('should return null if there are no radio items', () => {
    render(
      <RadioGroup
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={[]}
      />,
    );

    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });

  it('should return one item span and not a radio control', () => {
    render(
      <RadioGroup
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={[
          {
            value: 'test-label',
            label: 'test label',
          },
        ]}
      />,
    );

    expect(screen.queryByRole('list')).not.toBeInTheDocument();
    expect(screen.getByText(/test label/i)).toBeInTheDocument();
  });

  it('should show error message on error', () => {
    render(
      <RadioGroup
        errorMessage="error message"
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={[
          {
            value: 'test-label',
            label: 'test label',
          },
        ]}
      />,
    );

    expect(screen.getByText(/error message/i)).toBeInTheDocument();
  });

  it('should be accesible', async () => {
    const { container } = render(
      <RadioGroup
        name={RadioGroupFixture.name}
        onChange={RadioGroupFixture.onChange}
        options={RadioGroupFixture.options}
      />,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
