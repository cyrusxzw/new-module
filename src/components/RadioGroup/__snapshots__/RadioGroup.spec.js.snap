// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> should render base component correctly 1`] = `
<div>
  <ul
    class="base"
  >
    <li
      class="radio"
    >
      <label
        class="label"
        for="option-morbi-id-nulla"
      >
        <input
          aria-checked="true"
          checked=""
          class="input"
          data-test-ref="RADIO_GROUP_TEST_REF"
          id="option-morbi-id-nulla"
          name="radio-group"
          tabindex="0"
          type="radio"
          value="morbi-id-nulla"
        />
        <span
          class="pot dark"
        />
        <span
          class="labelContent dark"
        >
          Morbi id nulla
        </span>
      </label>
    </li>
    <li
      class="radio"
    >
      <label
        class="label"
        for="option-nullam-ultrices"
      >
        <input
          aria-checked="false"
          class="input"
          data-test-ref="RADIO_GROUP_TEST_REF"
          id="option-nullam-ultrices"
          name="radio-group"
          tabindex="0"
          type="radio"
          value="nullam-ultrices"
        />
        <span
          class="pot dark"
        />
        <span
          class="labelContent dark"
        >
          Nullam ultrices
        </span>
      </label>
    </li>
  </ul>
  <div
    class="errorMessage"
  >
    Test Error Messsage
  </div>
</div>
`;
